plc/CMEncrypt.1
.TH CMEncrypt 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
CMEncrypt - Send Encrypted Payload

.SH SYNOPSIS
.BR CMEncrypt
.RI [ options ] 
.RI [ device ] 
.RI [ device ] 
.RI [ ... ]

.SH DESCRIPTION
Transmit a file over powerline using CM_ENCRYPTED_PAYLOAD messages.
This program tests and demonstrates proper (or improper) message routing but does not support encrypted transmission.
All payloads are sent unencrypted at this time.

.SH OPTIONS

.TP
-\fBA\fI status\fR
The AVLN Status.
This indicates the current association status and capabilities of the sending station.
Permitted values are \fB0\fR through \fB255\fR but only value 0 through 5 and 8 are recognized by the firmware.
Values may be entered in decimal, hexadecimal or binary format.
Hexadecimal value are prefixed with "\fB0x\fR" or "\fB0X\fR".
Binary values are prefixed with "\fB0b\fR" or "\fB0B\fR".
The default is \fB0\fR.

.TP
-\fBf\fI file\fR
The name of the payload file.
File content may be text or binary and forms the payload of one or more messages.
Payloads are transmitted unencrypted when PID is 4 and PEKS is 15 which is default.

.TP
-\fBK \fIPEKS\fR
The Payload Encryption Key Select.
This value indicates the key used to encrypt the payload.
Permitted values are \fB0\fR through \fB15\fR.
Values may be entered in decimal, hexadecimal or binary format.
Hexadecimal value are prefixed with "\fB0x\fR" or "\fB0X\fR".
Binary values are prefixed with "\fB0b\fR" or "\fB0B\fR"..
The default is \fB15\fR for "No Encryption".

.TP
-\fBP \fIPID\fR
The Protocol Identifier.
This value detemines how the payload is interpreted by powerline devices.
Permitted values are \fB0\fR through \fB4\fR.
The default is \fB4\fR for "HLE Protocol".

.TP
-\fBv\fR
Print incoming and outgoing Ethernet frames in hex dump format on stdout.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ? ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
\fIdevice\fR
The Media Access Control (MAC) address of a remote host, not a powerline device.
Addresses are 6 hexadecimal octets optionally separated by colon.
For example, the addresses "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are all valid and equivalent.
For convenience, the symbolic address "\fBlocal\fR" resolves to "00:b0:52:00:00:01" and the symbolic addresses "\fBall\fR" and "\fBbroadcast\fR" both resolve to "ff:ff:ff:ff:ff:ff".

.SH INTERNALS
This program lets the user specify some CM_ENCRYPTED_PAYLOAD fields but not others.
The following are fields that the user cannot specify.
The program automatically computes these values.

.TP
.BR PRN
The Protocol Run Number.
This is a random number used to distinguish between runs having the same protocol.
This program uses the system time variable.

.TP
.BR UUID
The Universally Unique Identifier.
This program uses the left-most 16 bytes of the SHA256 digest for the payload.

.SH REFERENCES
Consult the \fIHomePlug AV Specification\fR or \fIAtheros Firmware Technical Reference Manual\fR for more information.

.SH EXAMPLES
The following example transmits file \fBmyfile.txt\fR to host \fB00:85:73:00:94:01\fR in unecrypted format because that is the program default.
The remote address is required and refers to the ultimate recipient, not the local device.

.PP
   # CMEncrypt -f myfile.txt 00:85:73:00:94:01
The next example does the same but specifies the program defaults.
Observe that operands can be expressed in decimal, binary or hexadecimal format.

.PP
   # CMEncrypt -f myfile.txt -A 0 -P 0b0100 -K 0x0F 00:85:73:00:94:01

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information is not available.
Qualcomm Atheros reserves the right to modify message structure or content in future firmware releases without any obligation to notify or compensate users of this program.

.SH SEE ALSO
None.

.SH CREDITS
 Charles Maier



CHANGES
open-plc-utils-0.0.6
---------------

1. Programs mac2pw and mac2pwd implement a new method of generating device passwords. The old method is supported but is not the default. 

open-plc-utils-0.0.5
---------------

open-plc-utils-0.0.4
---------------

open-plc-utils-0.0.3
---------------

1. Added header VisualStudioNET/include/inttype.h to define print specifier constant PRId64 for Windows systems.

2. Replaced conditional compiler code around print specifiers "lld" and "I64d" and replaced single specifier PRId64 as defined by the system.

3. Renamed man pages ending in .7 with .1 to reflect that fact that they will now be installed man1 instead of man7.

4. Replaced const HOMEPLUG_MTYPE with constant ETH_P_HPAV throughout.

5. Correct endianess of FIELD_ETH_TYPE in module coqos_add.c.

6. Updated chipset codes and implemented table-driven chipset detection algorithm.

7. Moved TLVNode structure definition from nvm/nvm.h to tools/tlv.h to prepare for exanded use of TLV structures in future MMEs.


open-plc-utils-0.0.2
---------------

1. Removed references to private vendor-specific MMEs

2. Removed unused archive targets in primary Makefile.

3. Added .gitignore to subdirectories where programs and other objects are compiled.

4. Added conditional __UCLIBC__ compile statement to ARCPPrint.c to account for variable argument definition error during compile; 

5. Added 'ignore' target to cascading Makefiles to re-write local .gitignore files.

6. Corrected email address for Florian Fainelli in module ARPCPrint.

7. Added new HomePlug AV SLAC Protocol MMEs definitions to homeplug.h.

8. Modified makefile 'ignore' target to use tr instead of sed because some sed implementations do not accept '\n'.



plc/CMEncrypt.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted (subject to the limitations
 *   in the disclaimer below) provided that the following conditions
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of
 *     its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   CMEncrypt - Send Encrypted Payload;
 *
 *   this program sends and receives raw ethernet frames and so needs
 *   root privileges; if you install it using "chmod 555" and "chown
 *   root:root" then you must login as root to run it; otherwise, you
 *   can install it using "chmod 4555" and "chown root:root" so that
 *   anyone can run it; the program will refuse to run until you get
 *   things right;
 *
 *
 *   Contributor(s):
 *      Charles Maier
 *
 *--------------------------------------------------------------------*/


/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <time.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../key/SHA256.h"
#include "../plc/plc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/hexencode.c"
#include "../tools/hexdecode.c"
#include "../tools/hexdump.c"
#include "../tools/todigit.c"
#include "../tools/error.c"
#include "../tools/synonym.c"
#endif

#ifndef MAKEFILE
#include "../plc/Confirm.c"
#include "../plc/Failure.c"
#include "../plc/Request.c"
#include "../plc/Devices.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#endif

#ifndef MAKEFILE
#include "../key/SHA256Reset.c"
#include "../key/SHA256Write.c"
#include "../key/SHA256Block.c"
#include "../key/SHA256Fetch.c"
#endif

#ifndef MAKEFILE
#include "../mme/EthernetHeader.c"
#include "../mme/HomePlugHeader.c"
#include "../mme/MMECode.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define CMENCRYPT_PEKS 0x0F
#define CMENCRYPT_AVLN 0x00
#define CMENCRYPT_PID 0x04

/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"A:f:i:K:P:qv",
		"device [device] [...]",
		"Send an encrypted payload using CM_ENCRYPTED_PAYLOAD",
		"A n\tAVLN Status [" LITERAL (CMENCRYPT_AVLN) "]",
		"K n\tPayload Encryption Key Select (PEKS) [" LITERAL (CMENCRYPT_PEKS) "]",
		"P n\tProtocol Identifier (PID) [" LITERAL (CMENCRYPT_PID) "]",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"f f\tpayload file",
		"q\tquiet mode",
		"v\tverbose mode",
		(char const *) (0)
	};

#ifndef __GNUC__
#pragma pack(push,1)
#endif

	struct __packed cm_encrypted_payload
	{
		uint8_t PEKS;
		uint8_t AVLN;
		uint8_t PID;
		uint16_t PRN;
		uint8_t PMN;
		uint8_t UUID [16];
		uint16_t LEN;
	}
	template =
	{
		CMENCRYPT_PEKS,
		CMENCRYPT_AVLN,
		CMENCRYPT_PID,
		0x0000,
		0x00,
		{
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00,
			0x00
		},
		0x0000
	};

#ifndef __GNUC__
#pragma pack (pop)
#endif

	struct _file_ file =
	{
		-1,
		(char const *) (0)
	};
	struct sha256 sha256;
	uint8_t digest [SHA256_DIGEST_LENGTH];
	time_t timer = time ((time_t *)(0));
	uint8_t packet [ETHER_MAX_LEN];
	uint8_t * buffer;
	signed extent;
	signed c;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'f':
			if ((file.file = open (file.name = optarg, O_BINARY|O_RDONLY)) == -1)
			{
				error (1, errno, "%s", file.name);
			}
			break;
		case 'P':
			template.PID = (byte)(uintspec (optarg, 0x00, 0x0F));
			break;
		case 'A':
			template.AVLN = (byte)(uintspec (optarg, 0x00, 0x08));
			break;
		case 'K':
			template.PEKS = (byte)(uintspec (optarg, 0x00, 0xFF));
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;

/*
 *	load entire file into memory;
 */

	if (file.file == -1)
	{
		error (1, ECANCELED, "No payload file given: Use -f <file>");
	}
	if ((extent = lseek (file.file, 0, SEEK_END)) == -1)
	{
		error (1, errno, FILE_CANTSIZE, file.name);
	}
	if (!(buffer = malloc (extent)))
	{
		error (1, errno, FILE_CANTLOAD, file.name);
	}
	if (lseek (file.file, 0, SEEK_SET))
	{
		error (1, errno, FILE_CANTHOME, file.name);
	}
	if (read (file.file, buffer, extent) != extent)
	{
		error (1, errno, FILE_CANTREAD, file.name);
	}
	close (file.file);
	SHA256Reset (&sha256);
	SHA256Write (&sha256, buffer, extent);
	SHA256Fetch (&sha256, digest);
	if (!argc)
	{
		error (1, ECANCELED, "No destination given");
	}
	openchannel (&channel);
	while ((argc) && (* argv))
	{
		signed offset = 0;
		signed remain = extent;

#if 0

		signed length = sizeof (struct packet_ms) - sizeof (template);

#else

		signed length = 502 - sizeof (template);

#endif

		if (!hexencode (channel.peer, sizeof (channel.peer), synonym (* argv, devices, SIZEOF (devices))))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		template.PRN = (uint16_t)(timer);
		template.PMN = 0;
		memcpy (template.UUID, digest, sizeof (template.UUID));
		while (remain)
		{
			uint8_t * memory = packet;
			if (length > remain)
			{
				length = remain;
			}
			template.PMN++;
			template.LEN = HTOLE16 (length);
			memset (memory, 0, sizeof (struct message));
			memory += EthernetHeader ((struct ethernet_hdr *)(memory), channel.peer, channel.host, channel.type);
			memory += HomePlugHeader ((struct homeplug_hdr *)(memory), 0, (CM_ENCRYPTED_PAYLOAD | MMTYPE_IND));
			memcpy (memory, &template, sizeof (template));
			memory += sizeof (template);
			memcpy (memory, buffer + offset, length);
			memory += length;
			extent = (signed)(memory - packet);
			if (extent < (ETHER_MIN_LEN - ETHER_CRC_LEN))
			{
				extent = (ETHER_MIN_LEN - ETHER_CRC_LEN);
			}
			if (sendpacket (&channel, packet, extent) < extent)
			{
				error (1, errno, CHANNEL_CANTSEND);
			}
			offset += length;
			remain -= length;
		}
		argc--;
		argv++;
	}
	closechannel (&channel);
	free (buffer);
	exit (0);
}

